plugins/sentinelone/komand_sentinelone/tasks/monitor_logs/task.py

```diff
@@ -0,0 +1,340 @@
+        }
+
+    def count_queries(self, queries: dict) -> int:
+        """
+        Count how many queries are to be made and return the value
+        """
+        return len(list(filter(lambda query: query is not None, queries.values())))
+
+    def log_pagination_cycle(self, cursors) -> None:
+        """
+        Provide logging information if task is in pagination cycle
+        """
+        pagination_info_string = "Pagination cursors received... "
+        for log_type, cursor in cursors.items():
+            if cursor is not None:
+                pagination_info_string += f" {log_type}:{cursor}"
+        self.logger.info("Paginating cycle in progress.")
+        self.logger.info(pagination_info_string)
+
+    def collect_logs_handler(
+        self,
+        queries: Dict,
+        cursors: Dict,
+        state: Dict,
+        lookback_timestamp: str,
+        current_run_timestamp: str,
+        total_forbidden_responses: int,
+        is_paginating: bool,
+    ) -> Tuple[List[Dict], int]:
+        """
+        Handle whether to query each endpoint based on input and pagination, gather returned data and error responses
+        """
+        all_logs = []
+        for log_type, check in queries.items():
+            if check:
+                cursor = cursors.get(log_type)
+                if is_paginating is False or (is_paginating is True and cursor):
+                    self.logger.info(f"Getting {log_type} between {lookback_timestamp} and {current_run_timestamp}...")
+                    if cursor:
+                        self.logger.info(f"Using next page cursor: {cursor}")
+                    logs, total_forbidden_responses = self.get_generic_logs(
+                        log_type,
+                        state,
+                        total_forbidden_responses,
+                        current_run_timestamp,
+                        lookback_timestamp,
+                        cursor,
+                    )
+                    self.logger.info(f"{len(logs)} logs found in {log_type} query")
+                    all_logs.extend(logs)
+                else:
+                    self.logger.info(f"Pagination in progress, skipping {log_type} collection...")
+        return all_logs, total_forbidden_responses
+
+    def get_generic_logs(
+        self,
+        log_type: str,
+        state: dict,
+        total_forbidden_responses: int,
+        current_run_timestamp: str,
+        last_run_timestamp: str = None,
+        pagination_cursor: str = None,
+    ) -> List[Dict]:
+        """
+        Determine log type and query that log set with query parameters determined by type, cursor, and last timestamp.
+        Set latest log timestamp and page cursor in state.
+        Return logs from query
+        """
+        timestamp_key = LOG_TYPE_MAP.get(log_type, {}).get("LAST_LOG_TIMESTAMP")
+        page_token_key = LOG_TYPE_MAP.get(log_type, {}).get("PAGE_TOKEN")
+        last_log_timestamp = state.get(timestamp_key)
```

Comment:
Sorry to bring it up again, but I think if we're storing the last timestamp for a a particular log type we should use it in the query? It should safeguard losing any data if their API is delayed on returning data to a certain time that we query from the last time we know we got data.

I can see the case that we would have `last_run_timestamp` and not the last log state - when a customer has log a and b checked, then decide they want log c. In that case I think we can revert to using the `last_run_timestamp` to get the query params?

This?
```
log_start_timestamp =  state.get(timestamp_key, last_run_timestamp)
query_params = self.get_query_params(log_type, log_start_timestamp, current_run_timestamp, pagination_cursor)
```

Reply:
This makes sense. On updating the code though we have to make sure that the last log timestamp falls within our cutoff (this will be ignored for pagination as we query on cursor).